pkg/logger/zap_otel.go

```diff
@@ -0,0 +1,266 @@
+	}
+	if o.zapCore != nil {
+		checked = o.zapCore.Check(entry, checked)
+	}
+	return checked
+}
+
+func (o OtelZapCore) Sync() error {
+	if o.zapCore != nil {
+		return o.zapCore.Sync()
+	}
+	// If no zap core is set, we don't need to sync anything
+	// as OpenTelemetry Core does not have a sync operation.
+	return nil
+}
+
+// WithLevel sets the minimum level for the OpenTelemetry Core log to be exported
+func WithLevel(level zapcore.Level) Option {
+	return Option(func(o *OtelZapCore) {
+		o.levelEnabler = level
+	})
+}
+
+// WithLevelEnabler sets the zapcore.LevelEnabler for determining which log levels to export
+func WithLevelEnabler(levelEnabler zapcore.LevelEnabler) Option {
+	return Option(func(o *OtelZapCore) {
+		o.levelEnabler = levelEnabler
+	})
+}
+
+// WithZapCore sets a zapcore.Core to be used for writing logs
+func WithZapCore(core zapcore.Core) Option {
+	return Option(func(o *OtelZapCore) {
+		o.zapCore = core
+	})
+}
+
+func (o OtelZapCore) Write(entry zapcore.Entry, fields []zapcore.Field) error {
+	var attributes []attribute.KeyValue
+	var spanCtx *oteltrace.SpanContext
+
+	// Add core-attached fields
+	for _, f := range o.fields {
+		if f.Key == "context" {
+			if ctxValue, ok := f.Interface.(oteltrace.SpanContext); ok {
+				spanCtx = &ctxValue
+			}
+		} else {
+			attributes = append(attributes, mapZapField(f))
+		}
+	}
+
+	// Add fields passed during log call
+	for _, f := range fields {
+		attributes = append(attributes, mapZapField(f))
+	}
+
+	// Add exception metadata
+	if entry.Level > zapcore.InfoLevel {
+		if entry.Caller.Defined {
+			attributes = append(attributes, semconv.ExceptionType(entry.Caller.String()))
+		}
+		if entry.Stack != "" {
+			attributes = append(attributes, semconv.ExceptionStacktrace(entry.Stack))
+		}
+	}
+
+	// Add span context attributes if available
+	if spanCtx != nil {
+		if spanCtx.TraceID().IsValid() {
+			attributes = append(attributes, attribute.String("trace_id", spanCtx.TraceID().String()))
+		}
+		if spanCtx.SpanID().IsValid() {
+			attributes = append(attributes, attribute.String("span_id", spanCtx.SpanID().String()))
+		}
+		attributes = append(attributes, attribute.String("trace_flags", spanCtx.TraceFlags().String()))
+	}
+
+	// Convert to OpenTelemetry KeyValues and emit
+	otelAttrs := make([]otellog.KeyValue, len(attributes))
+	for i, attr := range attributes {
+		otelAttrs[i] = otellog.KeyValue{
+			Key:   string(attr.Key),
+			Value: mapAttrValueToLogAttrValue(attr.Value),
+		}
+	}
+
+	logRecord := otellog.Record{}
+	logRecord.SetBody(otellog.StringValue(entry.Message))
+	logRecord.SetSeverity(mapZapSeverity(entry.Level))
+	logRecord.SetSeverityText(entry.Level.String())
+	logRecord.SetTimestamp(entry.Time)
+	logRecord.SetObservedTimestamp(time.Now())
+	logRecord.AddAttributes(otelAttrs...)
+
+	o.logger.Emit(context.Background(), logRecord)
+
+	// If a zapCore is set, delegate the write to Zap as well
+	if o.zapCore != nil {
+		return o.zapCore.Write(entry, fields)
+	}
```

Comment:
This seem really odd. Zap already has mechanisms for writing to multiple cores: https://pkg.go.dev/go.uber.org/zap/zapcore#NewTee

Layering like this doesn't make sense because they each own their own Enabled() and Check() implementations